apps/tap/test/tap.js

```diff
@@ -62,64 +85,328 @@ contract('Tap app', accounts => {
+
+        let balance = await tap.poolBalance(ETH, { from: authorized })
+        assert.equal(balance.toNumber(), INITIAL_ETH_BALANCE - tapAmt, 'pool balance should be decreased by tap amount')
+        assert.equal(await getBalance(vault.address), tapAmt, 'vault balance should be equal to tap amount')
+      })
+    })
+
+    context("ERC20", () => {
+      it("it should transfer a tap-defined amount of ERC20 from the collateral pool to the vault", async () => {
+        await tap.addTokenTap(token1.address, tapAmt, { from: authorized })
+        assert.equal(await token1.balanceOf(pool.address), 10000, 'pool should have ERC20 token balance')
+        await tap.withdraw(token1.address, { from: authorized })
+
+        assert.equal(await token1.balanceOf(pool.address), 10000 - tapAmt, 'token balance in pool should decrease by tap amount')
+        assert.equal(await token1.balanceOf(vault.address), tapAmt, 'vault should have updated token balance')
+      })
+    })
+
+    it("it should revert if sender does not have 'WITHDRAW_ROLE'", async () => {
+      return assertRevert(async () => {
+        await tap.withdraw(token1.address, { from: unauthorized })
+      })
+    })
   })
-})
 
-// contract("Tap", accounts => {
-//   context("initialize", () => {
-//     it("should initialize tap rate, collateral pool and vault", async () => {});
-//
-//     it("should revert on re-initialization", async () => {});
-//   });
-//
-//   context("withdraw", () => {
-//     context("ETH", () => {
-//       it("should transfer a tap-defined amount of ETH from the collateral pool to the vault", async () => {});
-//     });
-//
-//     context("ERC20", () => {
-//       it("should transfer a tap-defined amount of ERC20 from the collateral pool to the vault", async () => {});
-//     });
-//
-//     it("it should revert if sender does not have 'WITHDRAW_ROLE'", async () => {});
-//   });
-//
-//   context("updateTap", () => {
-//     it("should update tap rate", async () => {});
-//   });
-//
-//   context("updateVault", () => {
-//     it("should update vault address", async () => {});
-//   });
-//
-//   context("updateCollateralPool", () => {
-//     it("should update collateral pool address", async () => {});
-//   });
-// });
+  context("> addTokenTap", () => {
+    context('sender has ADD_TOKEN_TAP_ROLE', () => {
+      context('and token is ETH or ERC20', () => {
+        context('and token does not already exist in mapping', () => {
+          it('it should add tap token in mapping', async () => {
+            token2 = await TokenMock.new(authorized, 10000)
+            const token3 = await TokenMock.new(authorized, 10000)
+
+            await tap.addTokenTap(ETH, tapAmt, { from: authorized })
+            await tap.addTokenTap(token2.address, tapAmt, { from: authorized })
+            await tap.addTokenTap(token3.address, tapAmt, { from: authorized })
+            assert(await tap.getWithdrawalValue(ETH))
```

Comment:
Should we assert on each event that the correct address is being emitted? Had trouble finding an example for this